src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -338,6 +340,764 @@ void dfcc_instrumentt::instrument_function(
     function_pointer_contracts);
 }
 
+static bool is_loop_id_pragma(const irep_idt &pragma)
+{
+  return id2string(pragma).rfind(CONTRACT_PRAGMA_loop_id, 0) == 0;
+}
+
+static size_t get_loop_id_from_pragma(const irep_idt &pragma)
+{
+  INVARIANT(
+    is_loop_id_pragma(pragma), "Wrong usage of get_loop_id_from_pragma");
+  const auto pragma_s = id2string(pragma);
+  size_t last_index = pragma_s.find_last_not_of("0123456789");
+  return atoi(pragma_s.substr(last_index + 1).c_str());
+}
+
+size_t dfcc_instrumentt::get_loop_id_from_target(
+  const goto_programt::instructiont::targett target)
+{
+  INVARIANT(
+    is_loop_instruction(target), "Wrong usage of get_loop_id_from_target");
+  const auto &pragma = get_loop_id_pragma(target);
+
+  return get_loop_id_from_pragma(pragma);
+}
+
+bool dfcc_instrumentt::is_loop_instruction(
+  const goto_programt::instructiont::targett target)
+{
+  const auto &pragma = get_loop_id_pragma(target);
+  return pragma != empty_pragma;
+}
+
+void dfcc_instrumentt::tag_loop_instruction(
+  goto_programt::instructiont::targett &target,
+  const irep_idt &pragma,
+  const size_t loop_id)
+{
+  target->source_location_nonconst().add_pragma(
+    id2string(pragma) + ":" + std::to_string(loop_id));
+}
+
+irep_idt
+dfcc_instrumentt::get_loop_id_pragma(const goto_programt::targett &target)
+{
+  for(const auto &pragma : target->source_location().get_pragmas())
+  {
+    if(is_loop_id_pragma(pragma.first))
+      return pragma.first;
+  }
+  return empty_pragma;
+}
+
+void dfcc_instrumentt::remove_loop_id_pragma(goto_programt::targett &target)
+{
+  for(const auto &pragma : target->source_location().get_pragmas())
+  {
+    if(is_loop_id_pragma(pragma.first))
+    {
+      target->source_location_nonconst()
+        .get_named_sub()
+        .add(ID_pragma)
+        ->second.remove(pragma.first);
+      return;
+    }
+  }
+}
+
+void dfcc_instrumentt::compute_loop_nesting_graph(
```

Comment:
Could we move this method and the loop graph node type declaration to a `dfcc_loop_utilst` class in a separate file ?